src/privileged/sockets/api.js

```diff
@@ -0,0 +1,470 @@
+ChromeUtils.defineModuleGetter(this, "ExtensionCommon",
+                               "resource://gre/modules/ExtensionCommon.jsm");
+
+let tm = Cc["@mozilla.org/thread-manager;1"].getService();
+let socketService = Cc["@mozilla.org/network/socket-transport-service;1"]
+                      .getService(Ci.nsISocketTransportService);
+
+let gManager = {
+  serverSocketMap: new Map(),
+  sendingSocketMap: new Map(),
+  socketCount: 0,
+  dataListeners: new Set(),
+};
+
+this.sockets = class extends ExtensionAPI {
+  getAPI(context) {
+    return {
+      sockets: {
+        async createServerSocket() {
+          let serverSocket = Cc["@mozilla.org/network/server-socket;1"]
+                               .createInstance(Ci.nsIServerSocket);
+          serverSocket.init(-1, true, -1); // init with random port
+          gManager.socketCount++;
+          gManager.serverSocketMap.set(serverSocket.port, serverSocket);
+          return serverSocket.port;
+        },
+
+        startListening(id) {
+          if (!gManager.serverSocketMap.has(id)) {
+            return;
+          }
+          let socket = gManager.serverSocketMap.get(id);
+          socket.asyncListen({
+            onSocketAccepted: (sock, transport) => {
+              let inputStream = transport.openInputStream(0, 0, 0);
+              let socketListener = {
+                onInputStreamReady: () => {
+                  try {
+                    let count = inputStream.available();
+                  } catch (e) {
+                    // probably closed.
+                    return;
+                  }
+                  let bis = Cc["@mozilla.org/binaryinputstream;1"]
+                              .createInstance(Ci.nsIBinaryInputStream);
+                  bis.setInputStream(inputStream);
+                  let buff = bis.readByteArray(5); // 32 bit int followed by a char = 5 bytes
+                  let meta = bufferpack.unpack('>Lc', buff);
+                  let string = bis.readBytes(meta[0]);
+                  switch (meta[1]) {
+                    case 'j':
+                      string = JSON.parse(string);
+                    case 'n':
+                      gManager.dataListeners.forEach((listener) => {
```

Comment:
Also, I'm wondering about the `forEach`. Do we ever expect there to be more than one listener registered? Not saying we shouldn't support multiple listeners, but it seems like the current implementation will only ever register one (in socket.js).

Reply:
> We also need to fire the event listeners for json encoded data (case `j`)? You could do something like:
> 
> ```js
> if (meta[1] == 'j') {
>   string = JSON.parse(string);
> } else if (meta[1] != 'n')
>   <error>
> }
> gManager.dataListeners.forEach(...)
> ```

There's no `break` in `case 'j'` - both cases fire event listeners with 'j' additionally parsing the JSON first. Probably a bit hard to read! I don't mind changing it to what you suggested. :)

Reply:
> Also, I'm wondering about the `forEach`. Do we ever expect there to be more than one listener registered? Not saying we shouldn't support multiple listeners, but it seems like the current implementation will only ever register one (in socket.js).

Since this was written as a separate module/repo, I probably tried to be a bit supportive of "alternative" consumers of this code. Now that we're putting everything in one place, I think you're right and we can make some things more specific.

Reply:
I see you've left the forEach here, and I think your initial reasoning on doing so is fine. Resolving.